arch/x86_64/mcount-insn.c

```diff
@@ -367,6 +446,11 @@ int disasm_check_insns(struct mcount_disasm_engine *disasm,
 			break;
 	}
 
+	/* 
+	 * instructions in prologue could jump back to it, the func should be 
+	 * unsupported in this case.
+	 */
+	i = 0;
```

Comment:
This can be a separate fix.  Hmm.. is it really a problem?

Reply:
It's only a problem if branches in the prologue are relocated. I tried it with a function that has a 2 byte jmp 0xfc (2 bytes backward) in the prologue and the program crashed because the target insn of the jump was replaced by the instrumentation code.

Reply:
> The relocated jcc jumps to the address of the original instruction. If that target address falls in the prologue (that was overwritten), we may have a problem.

I don't think it's a problem, and we can simply ignore that instruction as it's PC-relative so the target address is still correct IMO.

Reply:
Oh it's PC-relative, indeed. But we patched the jcc to jump to a trampoline when condition is met. The trampoline jumps to an absolute address (the address of the code in the text segment).

Reply:
I thought you only patched offset in the jcc leaving the condition as is, no?

Reply:
Yes, the offset in the jcc was patched leaving the condition as is. But the new offset is the displacement between the jcc and the trampoline, not the old displacement between the jcc and it's original target.

Reply:
Right, then I think it's ok as rel-jump in prologue should go an insn in the prologue and see the same condition and then go to the trampoline (or not).  So we don't need to worry about jumps to prologue in prologue, right?

If it's not, I still think it should be a different issue so a different commit.

Reply:
What I'm concerning is when the **target** instruction (of the rel-jump in prologue) is modified.  We don't need to worry about the unmodified instructions IMO.  The modified case is only either LEA, CALL or JMP/Jcc (with this change).  And LEA, CALL and Jcc32 are fine since they can come at the last in the prologue so won't change start offsets of other instructions.  And I think Jcc8 is also fine since it keeps the instruction size.

Reply:
The disp of Jcc is not the only part of the problem. In this patch (and the patch that handles CALLs as well) The target of the trampoline is the target of the Jcc. If the target of the Jcc is the middle of the prologue we may have a problem. Here is an example with Jcc (Same can happen with JMPs and CALLs):

Let's say we want to instrument a function with 3 instructions in the prologue. This function has a Jcc8 that branches to the 2nd instruction.
 ```
Sym:
0x0: Insn1
0x2: Insn2
0x4: Jcc -4
 ```
The result is something similar to this:
 ```
Sym:
0x0: CALL dentry_trampoline
0x5: NOP
 ```
And the out of line execution buffer should be similar to this:
 ```
0x200: Insn1 
0x202: Insn2 
0x204: Jcc +10
0x206: Some padding
0x210: Indirect JMP 0x2   <-- Notice the 2nd instruction (Insn2) is not in 0x2 anymore 
  ```

Reply:
Thanks a lot for providing an example.  So my idea is just to keep `Jcc -4` unchanged.  We don't need the jump table in this case, right?

Reply:
Yes, in this case we don't need it. But we can't keep the same offset all the time. We need to patch it because if we relocate an instruction between the jcc and the target, its size may change. And the displacement of the jcc may be invalid.

Reply:
But as I mentioned, (currently) all the instructions that change the size is at least 5 byte so it cannot have Jcc after itself.

Reply:
Right, but we need to be careful later and take this into consideration.